Add `Item::from_def_id_and_kind` to reduce duplication in rustdoc

jyn514 opened the pull request:
This makes it harder to make typos, and also makes it much more clear what's intentionally different rather than a typo (look for `what_rustc_thinks`).

Found this while working on https://github.com/rust-lang/rust/issues/76998, I really didn't want to add `const_visibility` in 20 different places.

r? @GuillaumeGomez

petrochenkov commented on src/librustdoc/clean/types.rs:
```diff
@@ -113,6 +113,26 @@ impl Item {
     }
 
+    /// Convenience wrapper around [`Self::from_inner`] which converts `hir_id` to a [`DefId`]
+    pub fn from_hir(hir_id: hir::HirId, inner: ItemEnum, cx: &DocContext<'_>) -> Item {
```
```suggestion
    pub fn from_hir_id_and_kind(hir_id: hir::HirId, inner: ItemEnum, cx: &DocContext<'_>) -> Item {
```

petrochenkov replied:
The naming convention in rustc is `struct Item { kind: ItemKind, ... }`, it would be great if rustdoc followed it.

So I suggest to rename `ItemEnum` -> `ItemKind` and `Item::inner` -> `Item::kind` as well, perhaps even in this PR (but not necessarily).

petrochenkov commented on src/librustdoc/clean/types.rs:
```diff
@@ -113,6 +113,26 @@ impl Item {
+        Item::from_inner(cx.tcx.hir().local_def_id(hir_id).to_def_id(), inner, cx)
+    }
+
+    pub fn from_inner(def_id: DefId, inner: ItemEnum, cx: &DocContext<'_>) -> Item {
```
```suggestion
    pub fn from_def_id_and_kind(def_id: DefId, inner: ItemEnum, cx: &DocContext<'_>) -> Item {
```

jyn514 commented:
@petrochenkov do you understand this error?

```
error: internal compiler error: compiler/rustc_middle/src/ty/query/mod.rs:105:1: `tcx.visibility(DefId(0:4 ~ foo[8787]::Bar))` unsupported by its crate
query stack during panic:
#0 [visibility] computing visibility of `Struct`
end of query stack
```

I can change it to use the `hir::Visibility` instead, but do you know why it's necessary?

jyn514 commented:
It's coming from this code:
```rust
impl Clean<Item> for doctree::Struct<'_> {
    fn clean(&self, cx: &DocContext<'_>) -> Item {
        Item::from_hir_id_and_kind(
            self.id,
            StructItem(Struct {
                struct_type: self.struct_type,
                generics: self.generics.clean(cx),
                fields: self.fields.clean(cx),
                fields_stripped: false,
            }),
            cx,
        )
    }
}
```

jyn514 commented:
Hmm, this comes from `impl Default for Providers`. Maybe rustdoc forgot to call `providers::provide` somewhere?

jyn514 commented:
Looks like the crate that *should* provide this is https://github.com/rust-lang/rust/blob/d902752866cbbdb331e3cf28ff6bba86ab0f6c62/compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs#L204.

petrochenkov commented:
@jyn514 
>error: internal compiler error: compiler/rustc_middle/src/ty/query/mod.rs:105:1: `tcx.visibility(DefId(0:4 ~ foo[8787]::Bar))` unsupported by its crate

IIRC, the `visibility` query is only implemented for other crates, but not for the current crate (`LOCAL_CRATE`).
You could try implementing it for the local crate, it would be generally useful, I think.
Every time `Visibility::from_hir` is used, that new query could be used instead.
It's preferable to do that as a separate PR because there may be performance effects.

jyn514 commented:
>  You could try implementing it for the local crate, it would be generally useful, I think.

What would that look like? Would it be as simple as adding `cdata.get_visibility` as a local provider?

petrochenkov commented:
Search for `tables.visibility` in `compiler\rustc_metadata\src\rmeta\encoder.rs`, that's the logic necessary to get visibilities for `DefId`s.
All that logic needs to be gathered into a single function `fn visibility(def_id: DefId) -> ty::Visibility`, and that function need to be assigned to the local `visibility` query with
```rust
pub fn provide(providers: &mut Providers) {
    *providers = Providers { visibility, ..*providers };
}
```

jyn514 commented:
I tried this:
```rust
        visibility: |tcx, id| {
            let cdata = CStore::from_tcx(tcx).get_crate_data(id.krate);
            cdata.get_visibility(id.index)
        },
```
but that panicked when loading an item in the local crate:
```
thread 'rustc' panicked at 'Failed to get crate data for crate0', compiler/rustc_metadata/src/creader.rs:130:32
```
Is there another way to get the `root.tables` for the local crate?

petrochenkov commented:
@jyn514 
`cdata`, including `root.tables` represents other crates' metadata structures decoded from `rlib`s or `rmeta`s.
The currently built crate doesn't have any of that.
It has all the data kept in HIR, tcx, and other similar structures.
When searching for `tables.visibility` I meant looking how they are built and encoded in the current crate, so that other crates could decode them as `cdata`.

jyn514 commented:
Ok, https://github.com/rust-lang/rust/blob/f958e6c24615c4b1f1e77b0ee3d4f32383fb3617/compiler/rustc_privacy/src/lib.rs#L236-L239
looks about right. I'll try turning that into a query - currently it handles both local and cross-crate IDs, I think I can split the `Some(hir_id)` case into the query and have everywhere else just call `tcx.visibility(def_id)`. I'm not quite sure where to get back the `span` and `descr` though - I found `tcx.def_span`, but the current logic for `descr` for non-local IDs is https://github.com/rust-lang/rust/blob/f958e6c24615c4b1f1e77b0ee3d4f32383fb3617/compiler/rustc_privacy/src/lib.rs#L349
which is missing most of the cases from https://github.com/rust-lang/rust/blob/f958e6c24615c4b1f1e77b0ee3d4f32383fb3617/compiler/rustc_hir/src/hir.rs#L2403-L2410

Anyway that's mostly worrying about code duplication, let me make sure it actually works first.

jyn514 commented:
That gave
```
compiler/rustc_privacy/src/lib.rs:346:17: unexpected node kind: Field(StructField { ... })
```
I'm not sure where that case is handled currently.

bors commented:
:umbrella: The latest upstream changes (presumably #77926) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

petrochenkov commented:
https://github.com/rust-lang/rust/pull/78077 has landed.

jyn514 commented:
Rebased and updated. This is still panicking on most tests:
```
error: internal compiler error: compiler/rustc_middle/src/hir/map/mod.rs:477:41: couldn't find hir id HirId { owner: DefId(0:0 ~ union[8787]), local_id: 0 } in the HIR map
```
Backtrace:
```
   8: rustc_middle::util::bug::bug_fmt
             at ./compiler/rustc_middle/src/util/bug.rs:14:5
   9: rustc_middle::hir::map::Map::get::{{closure}}
             at ./compiler/rustc_middle/src/hir/map/mod.rs:477:41
  10: core::option::Option<T>::unwrap_or_else
             at ./library/core/src/option.rs:427:21
  11: rustc_middle::hir::map::Map::get
             at ./compiler/rustc_middle/src/hir/map/mod.rs:477:9
  12: rustc_middle::ty::<impl rustc_middle::ty::context::TyCtxt>::opt_item_name::{{closure}}
             at ./compiler/rustc_middle/src/ty/mod.rs:2796:32
  13: core::option::Option<T>::and_then
             at ./library/core/src/option.rs:662:24
  14: rustc_middle::ty::<impl rustc_middle::ty::context::TyCtxt>::opt_item_name
             at ./compiler/rustc_middle/src/ty/mod.rs:2794:9
  15: rustdoc::clean::types::Item::from_def_id_and_kind
             at ./src/librustdoc/clean/types.rs:127:19
  16: rustdoc::clean::types::Item::from_hir_id_and_kind
             at ./src/librustdoc/clean/types.rs:118:9
  17: <rustdoc::doctree::Module as rustdoc::clean::Clean<rustdoc::clean::types::Item>>::clean
             at ./src/librustdoc/clean/mod.rs:265:33
  18: rustdoc::clean::utils::krate
             at ./src/librustdoc/clean/utils.rs:44:22
```
I think the issue is that `opt_item_name` doesn't check for the crate root, only `item_name`:
https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/ty/mod.rs#L2793-L2797
https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/ty/mod.rs#L2919-L2934
I'll try rewriting `item_name` in terms of `opt_item_name` and see if that helps.

jyn514 commented:
This is now failing a variety of different tests:

<details><summary>Failures</summary>

```
failures:
    [rustdoc] rustdoc/assoc-types.rs
    [rustdoc] rustdoc/check-styled-link.rs
    [rustdoc] rustdoc/default-trait-method-link.rs
    [rustdoc] rustdoc/external-macro-src.rs
    [rustdoc] rustdoc/foreigntype-reexport.rs
    [rustdoc] rustdoc/hidden-impls.rs
    [rustdoc] rustdoc/impl-disambiguation.rs
    [rustdoc] rustdoc/inline_cross/macro-vis.rs
    [rustdoc] rustdoc/inline_cross/renamed-via-module.rs
    [rustdoc] rustdoc/inline_local/macro_by_example.rs
    [rustdoc] rustdoc/intra-doc-crate/additional_doc.rs
    [rustdoc] rustdoc/intra-doc-crate/hidden.rs
    [rustdoc] rustdoc/intra-doc-crate/submodule-outer.rs
    [rustdoc] rustdoc/intra-doc-link-enum-struct-field.rs
    [rustdoc] rustdoc/intra-doc-link-mod-ambiguity.rs
    [rustdoc] rustdoc/intra-doc-link-private.rs
    [rustdoc] rustdoc/intra-link-associated-defaults.rs
    [rustdoc] rustdoc/intra-link-associated-items.rs
    [rustdoc] rustdoc/intra-link-disambiguators-removed.rs
    [rustdoc] rustdoc/intra-link-prim-precedence.rs
    [rustdoc] rustdoc/intra-link-proc-macro.rs
    [rustdoc] rustdoc/intra-link-reexport-additional-docs.rs
    [rustdoc] rustdoc/intra-link-self.rs
    [rustdoc] rustdoc/intra-link-trait-impl.rs
    [rustdoc] rustdoc/intra-link-trait-item.rs
    [rustdoc] rustdoc/intra-links-anchors.rs
    [rustdoc] rustdoc/intra-links.rs
    [rustdoc] rustdoc/issue-28478.rs
    [rustdoc] rustdoc/issue-32395.rs
    [rustdoc] rustdoc/issue-34473.rs
    [rustdoc] rustdoc/issue-55364.rs
    [rustdoc] rustdoc/issue-72340.rs
    [rustdoc] rustdoc/link-assoc-const.rs
    [rustdoc] rustdoc/primitive-reexport.rs
    [rustdoc] rustdoc/proc-macro.rs
    [rustdoc] rustdoc/pub-restricted.rs
    [rustdoc] rustdoc/redirect-const.rs
    [rustdoc] rustdoc/redirect-rename.rs
    [rustdoc] rustdoc/redirect.rs
    [rustdoc] rustdoc/src-links-external.rs
    [rustdoc] rustdoc/struct-field.rs
    [rustdoc] rustdoc/synthetic_auto/complex.rs

test result: FAILED. 116 passed; 42 failed; 229 ignored; 0 measured; 0 filtered out
```

</details>

I don't have time to go through them all right now, I'll try to come back to it this weekend.

jyn514 commented:
One of the failures, which possibly is causing others, is that rustdoc is no longer generating the correct URLs.

- Before: `../assoc_types/trait.Index.html#associatedtype.Output`
- After: `../assoc_types/assoc_types/trait.Index.html#associatedtype.Output`

```
---- [rustdoc] rustdoc/assoc-types.rs stdout ----

error: htmldocck failed!
status: exit code: 1
command: "/usr/bin/python" "/home/joshua/rustc/src/etc/htmldocck.py" "/home/joshua/rustc/build/x86_64-unknown-linux-gnu/test/rustdoc/assoc-types" "/home/joshua/rustc/src/test/rustdoc/assoc-types.rs"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
18: @has check failed
	`XPATH PATTERN` did not match
	// @has - '//*[@class="rust fn"]//a[@href="../assoc_types/trait.Index.html#associatedtype.Output"]' 'Output'
```

```rust
$ cat /home/joshua/rustc/src/test/rustdoc/assoc-types.rs
// ignore-tidy-linelength

#![crate_type="lib"]

// @has assoc_types/trait.Index.html
pub trait Index<I: ?Sized> {
    // @has - '//*[@id="associatedtype.Output"]//code' 'type Output: ?Sized'
    type Output: ?Sized;
    // @has - '//*[@id="tymethod.index"]//code' \
    //      "fn index<'a>(&'a self, index: I) -> &'a Self::Output"
    // @has - '//*[@id="tymethod.index"]//code//a[@href="../assoc_types/trait.Index.html#associatedtype.Output"]' \
    //      "Output"
    fn index<'a>(&'a self, index: I) -> &'a Self::Output;
}

// @has assoc_types/fn.use_output.html
// @has - '//*[@class="rust fn"]' '-> &T::Output'
// @has - '//*[@class="rust fn"]//a[@href="../assoc_types/trait.Index.html#associatedtype.Output"]' 'Output'
pub fn use_output<T: Index<usize>>(obj: &T, index: usize) -> &T::Output {
    obj.index(index)
}
```

jyn514 commented:
@Nemo157 helped me put together a list of what's breaking (view with `less -R differences.log`). I'll see if I can add this natively to rustbuild sometime soon.

[differences.log](https://github.com/rust-lang/rust/files/5490490/differences.log)

jyn514 commented:
`pub-restricted` is failing because the HIR visibility doesn't match `tcx.visibility`:

```
// this is the HIR visibility
2:rustcDEBUG rustdoc::clean Restricted { path: Path { span: src/test/rustdoc/pub-restricted.rs:16:5: 16:9 (#0), res: Def(Mod, DefId(0:0 ~ foo[8787])), segments: [PathSegment { ident: self#0, hir_id: Some(HirId { owner: DefId(0:9 ~ foo[8787]::FooSelf), local_id: 2 }), res: Some(Err), args: None, infer_args: false }] }, hir_id: HirId { owner: DefId(0:9 ~ foo[8787]::FooSelf), local_id: 1 } } FooSelf
// this is tcx.visibility()
2:rustcDEBUG rustdoc::clean::types name=Some("FooSelf"), def_id=DefId(0:9 ~ foo[8787]::FooSelf), visibility=Inherited
```

I found it's because `impl Clean for ty::Visibility` is wrong and completely ignores restricted paths: https://github.com/rust-lang/rust/blob/8752a560b9f50e04c38ab75e2b5c37d03d3214d9/src/librustdoc/clean/mod.rs#L1825

But when I try to implement it, rustdoc wants more data than I have available:

```rust
    fn clean(&self, cx: &DocContext<'_>) -> Visibility {
        match *self {
            ty::Visibility::Public => Visibility::Public,
            ty::Visibility::Restricted(module) => {
                if module.is_top_level_module() {
                    Visibility::Crate
                } else {
                    Visibility::Restricted(module, cx.tcx.def_path(module).clean(cx))
                }
            }
            ty::Visibility::Invisible => Visibility::Inherited,
        }
    }
```
```
error[E0599]: no method named `clean` found for struct `DefPath` in the current scope
    --> src/librustdoc/clean/mod.rs:1767:76
     |
1767 |                     Visibility::Restricted(module, cx.tcx.def_path(module).clean(cx))
     |                                                                            ^^^^^ method not found in `DefPath`
```

and when I try to add that impls, there's not enough info left that's been serialized:

```rust
impl Clean<Path> for hir::definitions::DefPath {
    fn clean(&self, cx: &DocContext<'_>) -> Path {
        Path {
            global: unimplemented!(),
            res: unimplemented!(),
            segments: unimplemented!(),
        }
    }
}
```

Is there a way to do this? Or should I just give up on using `visibility()` and pass it in from the HIR everywhere it's used?

jyn514 commented:
The remaining test failure is that `tcx.visibility()` returns `Visibility::Public` on enum variants, not `Visibility::Inherited`. @petrochenkov is that intentional? The `hir::Visibility` is `Inherited`, I'm having to override the visibility manually.

petrochenkov commented:
@jyn514 
>The remaining test failure is that `tcx.visibility()` returns `Visibility::Public` on enum variants, not `Visibility::Inherited`.

`tcx.visibility()` returns semantic visibility (`ty::Visibility`), it doesn't have an `Inherited` variant as you can see, `Inherited` is replaced with the actual visibility of the item from which it is inherited from.
Variants inherit visibility from their enums, so if the enum is public, then its variants are public as well.

What happens in rustdoc doesn't override visibility on variants?
Does it print `enum E { pub V }`? Then that's a rustdoc's issue, it shouldn't print `pub` on variants regardless of their visibility.

jyn514 commented:
> What happens in rustdoc doesn't override visibility on variants? Does it print enum E { pub V }? Then that's a rustdoc's issue, it shouldn't print pub on variants regardless of their visibility.

Ok, I fixed this on rustdoc's end in https://github.com/rust-lang/rust/pull/77820/commits/946d536b46e4af2f7ad498dde1f89245fd82dff9.

This is finally working :tada: :tada: It does have a change in behavior: now that visibilities come from `tcx.visibility`, they use `ty::Visibility` instead of `hir::Visibility`, so `pub(self)` appears as `pub(in a::b)` instead. I'm not sure how to fix this without ignoring `tcx.visibility` altogether.

petrochenkov commented on src/librustdoc/clean/inline.rs:
```diff
@@ -124,16 +124,8 @@ crate fn try_inline(
-        def_id: did,
-    });
+    let what_rustc_thinks = clean::Item::from_def_id_and_parts(did, Some(name), inner, cx);
+    ret.push(clean::Item { attrs, visibility: clean::Public, ..what_rustc_thinks });
```
Why does rustdoc override visibility here?

jyn514 replied:
Oops, this is left over from trying to fix the `enum E { pub V }` bug. I reverted it.

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1774,33 +1724,25 @@ impl<'tcx> Clean<Constant> for ty::Const<'tcx> {
+            StructFieldItem(self.ty.clean(cx)),
+            cx,
+        );
+        Item { visibility: self.vis.clean(cx), ..what_rustc_thinks }
```
Why does rustdoc override visibility here?

jyn514 replied:
It turns out this is actually necessary: if you have

```rust
pub enum Foo {
    Bar {
        qux: (),
    }
}
```
rustdoc will render the `Bar` as `pub(crate) Bar` unless you change the visibility: https://github.com/rust-lang/rust/pull/77820/checks?check_run_id=1400308817#step:24:1755.

I added a comment and changed it back to the old behavior.

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1774,33 +1724,25 @@ impl<'tcx> Clean<Constant> for ty::Const<'tcx> {
+            StructFieldItem(cx.tcx.type_of(self.did).clean(cx)),
+            cx,
+        );
+        Item { visibility: self.vis.clean(cx), ..what_rustc_thinks }
```
Why does rustdoc override visibility here?

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1813,55 +1755,57 @@ impl Clean<Visibility> for hir::Visibility<'_> {
+                }
+            }
+            ty::Visibility::Invisible => Visibility::Inherited,
+        }
```
You can use the logic from

https://github.com/rust-lang/rust/blob/master/compiler/rustc_privacy/src/lib.rs#L1741-L1753

to prefer "" to "pub(crate)"/"pub(in path)", and "pub(crate)" to "pub(in path)".

petrochenkov replied:
It can be applied to `impl Clean<Visibility> for hir::Visibility<'_> { ... }` as well since you are using def-paths now anyway.

jyn514 replied:
That works on the HIR, so I need a way to get the parent module of a DefId. I have some code for that in https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/librustdoc/passes/collect_intra_doc_links.rs#L810-L828, do you think it makes sense to turn that into a query? Or should I just make a free function in rustdoc?

petrochenkov replied:
There's a readily available method for obtaining a parent module of `HirId` - `tcx.parent_module(hir_id)` (and you don't need this functionality for other crates).

jyn514 replied:
> and you don't need this functionality for other crates

Hmm, I think there's a bug in rustc somewhere then: [`impl MetadataExt for Metadata`](https://doc.rust-lang.org/nightly/src/std/os/linux/fs.rs.html#327) has a `Visibility::Restricted` even though it's from another crate.

```
DEBUG rustdoc::clean::inline build_impl: impl Some(DefId(1:7424 ~ std[7f8a]::os::linux::fs::MetadataExt)) for Some(DefId(1:7039 ~ std[7f8a]::fs::Metadata))
DEBUG rustdoc::clean::types name=None, def_id=DefId(1:7442 ~ std[7f8a]::os::linux::fs::{impl#0})
thread 'rustc' panicked at 'only local items can have restricted visibility (while calculating vis of DefId(1:7420 ~ std[7f8a]::os::linux::fs))', src/librustdoc/clean/mod.rs:1762:51
```

This is from local changes so the backtrace doesn't match the PR - let me know if I should push the broken code.

<details><summary>Backtrace</summary>

```
stack backtrace:
   0: rust_begin_unwind
   1: std::panicking::begin_panic_fmt
   2: rustdoc::clean::ty_visibility::{{closure}}
   3: rustdoc::clean::ty_visibility
   4: rustdoc::clean::types::Item::from_def_id_and_parts
   5: rustdoc::clean::inline::build_impl
   6: rustc_session::utils::<impl rustc_session::session::Session>::time
   7: rustdoc::passes::collect_trait_impls::collect_trait_impls
   8: rustc_session::utils::<impl rustc_session::session::Session>::time
   9: rustdoc::core::run_global_ctxt
  10: rustc_interface::passes::QueryContext::enter
  11: rustc_session::utils::<impl rustc_session::session::Session>::time
  12: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  13: rustc_span::with_source_map
  14: rustc_interface::interface::create_compiler_and_run
  15: rustdoc::core::run_core
  16: rustdoc::main_options
  17: scoped_tls::ScopedKey<T>::set
  18: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

The `def_id` is coming from https://github.com/rust-lang/rust/blob/master/src/librustdoc/clean/inline.rs#L341.
```

</details>

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1877,35 +1821,29 @@ impl Clean<VariantStruct> for rustc_hir::VariantData<'_> {
+            VariantItem(Variant { kind: self.def.clean(cx) }),
+            cx,
+        );
+        // don't show `pub` for variants, which are always public
```
This should probably go into rendering of `VariantItem` instead.

jyn514 replied:
I would prefer not to do that, at least not in this PR. Right now the rendering is handled here: https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/librustdoc/html/render/mod.rs#L3114

`render_struct` takes a `clean::Item` so it can be used for both variant structs and regular struct items: https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/librustdoc/html/render/mod.rs#L3251-L3259

If I changed this to have the 'real' visibility I'd have to add more state to `render` to tell the difference between a variant and a struct. And I'm not sure this is the only place `VariantItem` is used, there may be others depending on this behavior.

jyn514 replied:
Actually it looks like `structhead` is the state I need already. But I'm still not sure this is the only place `VariantItem` is used.

petrochenkov replied:
Ok.

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1935,16 +1873,14 @@ impl Clean<Item> for ty::VariantDef {
+            VariantItem(Variant { kind }),
+            cx,
+        );
+        // don't show `pub` for fields, which are always public
```
This should probably go into rendering of `VariantItem` instead.

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1926,7 +1864,7 @@ impl Clean<Item> for ty::VariantDef {
                         name: Some(field.ident.name.clean(cx)),
                         attrs: cx.tcx.get_attrs(field.did).clean(cx),
-                        visibility: field.vis.clean(cx),
+                        visibility: Visibility::Inherited,
```
This should probably go into rendering of `VariantStruct` instead.

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -2356,57 +2276,41 @@ impl Clean<Item> for doctree::ForeignItem<'_> {
-        }
+            cx,
+        );
+        // need to override attrs explicitly in case this was inlined
```
Not sure what this means.

jyn514 replied:
Here are the errors if the attributes aren't overriden:

```
---- [rustdoc] rustdoc/inline_local/macro_by_example.rs stdout ----
12: @has check failed
	`XPATH PATTERN` did not match
	    // @has - '//*[@class="docblock"]' 'docs for foo'

---- [rustdoc] rustdoc/intra-link-disambiguators-removed.rs stdout ----
44: @has check failed
	`XPATH PATTERN` did not match
	// @has - '//a[@href="../intra_link_disambiguators_removed/fn.Name.html"]' "fn@Name"
48: @has check failed
	`XPATH PATTERN` did not match
	// @has - '//a[@href="../intra_link_disambiguators_removed/trait.Name.html"]' "trait@Name"
```

The comment is wrong, it doesn't have anything to do with inlining. The issue is that rustc_metadata strips the attributes on macros: https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/test/rustdoc/intra-link-disambiguators-removed.rs#L42-L49

So this needs to get the attributes from the HIR instead.

petrochenkov replied:
>The issue is that rustc_metadata strips the attributes on macros

Wait, what.
It shouldn't do that.
`fn encode_info_for_macro_def` writes attributes into metadata, so something else is happening here.

jyn514 replied:
Hmm, I'm not sure. Rustdoc isn't doing anything special with the attributes:

https://github.com/rust-lang/rust/blob/75042566d1c90d912f22e4db43b6d3af98447986/src/librustdoc/visit_ast.rs#L622-L625

`def` here is `hir::MacroDef`.

jyn514 replied:
The issue was that `hir().attrs()` was ignoring attributes on `MacroDef`. https://github.com/rust-lang/rust/pull/77820/commits/b9f769045e2e04cd361e01c3b67851814a2e4d04

petrochenkov commented on src/librustdoc/clean/types.rs:
```diff
@@ -113,6 +113,50 @@ impl Item {
         self.attrs.doc_value()
     }
 
+    /// Convenience wrapper around [`Self::from_inner`] which converts `hir_id` to a [`DefId`]
```
```suggestion
    /// Convenience wrapper around [`Self::from_def_id_and_parts`] which converts `hir_id` to a [`DefId`]
```

petrochenkov commented on src/librustdoc/clean/types.rs:
```diff
@@ -113,6 +113,50 @@ impl Item {
+
+        debug!("name={:?}, def_id={:?}", name, def_id);
+
+        // `span_if_local()` lies about functions and only gives the span of the function signature
```
Looks like you are re-implementing `fn span_with_body` here.

jyn514 replied:
I didn't know that existed, thanks!

petrochenkov commented on src/librustdoc/html/format.rs:
```diff
@@ -1089,19 +1089,32 @@ impl Function<'_> {
             clean::Inherited => Ok(()),
             clean::Visibility::Crate => write!(f, "pub(crate) "),
+            // If this is `pub(crate)`, `path` will be empty.
+            clean::Visibility::Restricted(did, _) if did.index == CRATE_DEF_INDEX => {
```
Looks like the logic here and in "`Clean for Visibility`" is duplicated.
Normalization "pub (in path)" -> ""/"path(crate)", and "pub(crate)" -> "" should either be done entirely here (perferable?) or entirely there.

jyn514 replied:
> Normalization "pub (in path)" -> ""/"path(crate)", and "pub(crate)" -> "" should either be done entirely here (perferable?)

It would be nice to do it here, but there's not enough context left: `print_with_space` has neither a `tcx` nor the `HirId` of the original item, so it can't call `tcx.parent_module(hir_id)` to know whether to display the visibility as `pub(self)` or not. I'll add this in `Clean for ty::Visibility` and `Clean for hir::Visibility` instead, moving the previous `"self"/"super"` logic to `hir::Visibility`.

jyn514 replied:
Actually, this information isn't available there either :/ I have just the visibility and not the `Item` it's being calculated for. I'll try turning both of those into a free function, maybe?

petrochenkov commented:
The PR name also needs an update.

jyn514 commented:
I think I addressed all the review comments. I'm waiting on guidance about https://github.com/rust-lang/rust/pull/77820#discussion_r523460494 before simplifying the rendering of `pub(in path)`.

petrochenkov commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1809,59 +1753,60 @@ impl Clean<Visibility> for hir::Visibility<'_> {
         match self.node {
```
I think you can actually remove `impl Clean<Visibility> for hir::Visibility<'_>` entirely?
It's now used only for resetting visibilities to inherited, so `visibility: Visibility::Inherited` can be used everywhere instead of `visibility: self.vis.clean(cx)`.

jyn514 replied:
This is also used for visibilities on imports, which are not always inherited (e.g. if the import is being inlined): https://github.com/rust-lang/rust/blob/fff92031402e88256dbd828dcb2f00bcd2f03df2/src/librustdoc/clean/mod.rs#L2243

That can't use `from_def_id_and_kind` because it makes up a fake DefId. There are also a few other places that aren't using `Inherited`:

https://github.com/rust-lang/rust/blob/fff92031402e88256dbd828dcb2f00bcd2f03df2/src/librustdoc/clean/mod.rs#L2096
https://github.com/rust-lang/rust/blob/fff92031402e88256dbd828dcb2f00bcd2f03df2/src/librustdoc/clean/mod.rs#L2152
https://github.com/rust-lang/rust/blob/fff92031402e88256dbd828dcb2f00bcd2f03df2/src/librustdoc/clean/mod.rs#L2223

bors commented:
:umbrella: The latest upstream changes (presumably #79070) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

petrochenkov commented:
r=me after rebasing and squashing commits.

https://github.com/rust-lang/rust/pull/77820#discussion_r523402990 can be investigated some time later.

jyn514 commented:
cc @rust-lang/rustdoc - this now displays `pub(self)` or `pub(super)` as `pub(in a::b)` instead, which is some sense a regression. This only shows up with `--document-private-items` and the paths are correct, they're just not as friendly. I plan to fix it in a follow-up, but let me know if I should fix it now instead.

GuillaumeGomez commented:
From the diff, the information remains the same. What changes is that it goes from "relative" to "absolute". So to sum things up:

I'm not too sure about the doc reading side. For example when you use `super`, you refer to the parent of the current scope whereas we now give a full path, which might make reading a bit less convenient (because it could be harder to figure out where you are when path isn't relative to your "current location"). However, this is more of a point of view rather than an absolute truth and I'd really like to know what @Manishearth and @ollie27 think about this.

bors commented:
:umbrella: The latest upstream changes (presumably #79104) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Manishearth commented:
> they're just not as friendly. I plan to fix it in a follow-up, but let me know if I should fix it now instead.

Followup is fine by me. I do think it should be fixed.

jyn514 commented:
@Manishearth FYI - Guillaume and I discussed offline that there's no way to get precisely the previous behavior back. I can recover `pub(self)` (in theory, I need to work out exactly how) and `pub(super)`, but the previous behavior with `super::a::super` will *always* be normalized to `pub(super)` now - the information about how it was written just isn't present in the metadata, only in the HIR.

Manishearth commented:
Yeah that's fine, I care about self and super only, the others are rare enough

jyn514 commented:
@bors r=petrochenkov rollup=never

I want to be able to bisect if this causes regressions.

bors commented:
:pushpin: Commit 46a99790cdf4dc633fe8b295b506b15c7ad72f1c has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 46a99790cdf4dc633fe8b295b506b15c7ad72f1c -->
<!-- homu: {"type":"Approved","sha":"46a99790cdf4dc633fe8b295b506b15c7ad72f1c","approver":"petrochenkov"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #79138) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

jyn514 commented:
@bors r=petrochenkov rollup=never

bors commented:
:pushpin: Commit 0e1a302f599ea1714f1ffa1a34dfb71ce2682c4c has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 0e1a302f599ea1714f1ffa1a34dfb71ce2682c4c -->
<!-- homu: {"type":"Approved","sha":"0e1a302f599ea1714f1ffa1a34dfb71ce2682c4c","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 0e1a302f599ea1714f1ffa1a34dfb71ce2682c4c with merge c4f836ad1aceb83507810d9499f56988fd24578d...
<!-- homu: {"type":"BuildStarted","head_sha":"0e1a302f599ea1714f1ffa1a34dfb71ce2682c4c","merge_sha":"c4f836ad1aceb83507810d9499f56988fd24578d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1416549848)
Approved by: petrochenkov
Pushing c4f836ad1aceb83507810d9499f56988fd24578d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1416549848"},"merge_sha":"c4f836ad1aceb83507810d9499f56988fd24578d"} -->

jyn514 commented:
@petrochenkov thank you so much for all your help, I wouldn't have been able to do this without you :)
